File: lcof-python/ListNode/5_合并k个已经排序链表.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 
# @param lists ListNode类一维数组 
# @return ListNode类
#
class Solution:
    # 归并排序的分治
    def mergeKLists(self , lists: List[ListNode]) -> ListNode:
        # write code here
        if len(lists) <= 1:
            if len(lists) == 1:
                return lists[0]
            else:
                return None
        iMid = len(lists) // 2
        # 分治
        ListsLeft = self.mergeKLists(lists[:iMid])
        ListsRight = self.mergeKLists(lists[iMid:])
        # 左右两边是否为空
        if not ListsLeft:
            return ListsRight
        if not ListsRight:
            return ListsLeft
        # 获取头节点
        if ListsLeft.val <= ListsRight.val:
            temp = ListsLeft
            ListsLeft = ListsLeft.next
        else:
            temp = ListsRight
            ListsRight = ListsRight.next
        head = temp

        while ListsLeft and ListsRight:
            if ListsLeft.val <= ListsRight.val:
                temp.next = ListsLeft
                ListsLeft = ListsLeft.next
            else:
                temp.next = ListsRight
                ListsRight = ListsRight.next
            temp = temp.next

        if ListsLeft:
            temp.next = ListsLeft
        else:
            temp.next = ListsRight
        
        return head

File: lcof-cpp/offer/1_ListNode/5_合并k个已经排序链表.cpp
/**
 * struct ListNode {
 *  int val;
 *  struct ListNode *next;
 *  ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
#include <vector>
class Solution
{
public:
    /**
     * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
     *
     *
     * @param lists ListNode类vector
     * @return ListNode类
     */

    ListNode *Merge2(ListNode *pHead1, ListNode *pHead2)
    {
        if (pHead1 == nullptr)
            return pHead2;
        if (pHead2 == nullptr)
            return pHead1;
        auto *head = new ListNode(-1);
        ListNode *cur = head;
        while (pHead1 && pHead2)
        {
            if (pHead1->val <= pHead2->val)
            {
                cur->next = pHead1;
                pHead1 = pHead1->next;
            }
            else
            {
                cur->next = pHead2;
                pHead2 = pHead2->next;
            }
            cur = cur->next;
        }
        if (pHead1)
            cur->next = pHead1;
        else
            cur->next = pHead2;
        return head->next;
    }

    ListNode *divideMerge(vector<ListNode *> &lists, int left, int right)
    {
        if (left > right)
            return nullptr;
        else if (left == right)
            return lists[left];
        int mid = (left + right) / 2;
        return Merge2(divideMerge(lists, left, mid), divideMerge(lists, mid + 1,
                                                                 right));
    }

    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        return divideMerge(lists, 0, lists.size() - 1);
    }
};